ipatests/test_ipaserver/test_kadmin.py

```diff
@@ -0,0 +1,125 @@
+#
+# Copyright (C) 2016 FreeIPA Contributors see COPYING for license
+#
+
+"""
+Test suite for creating principals via kadmin.local and modifying their keys
+"""
+
+import os
+import pytest
+import tempfile
+
+from ipalib import api
+
+from ipaserver.install import installutils
+
+
+@pytest.yield_fixture()
+def keytab():
+    fd, keytab_path = tempfile.mkstemp(suffix='.keytab')
+    os.close(fd)
+
+    try:
+        yield keytab_path
+    finally:
+        try:
+            os.remove(keytab_path)
+        except OSError:
+            pass
+
+
+@pytest.fixture()
+def service_in_kerberos_subtree(request):
+    princ = u'svc1/{0.host}@{0.realm}'.format(api.env)
+    installutils.kadmin_addprinc(princ)
+
+    def fin():
+        try:
+            installutils.kadmin(
+                'delprinc -force {}'.format(princ))
+        except Exception:
+            pass
+    request.addfinalizer(fin)
+    return princ
+
+
+@pytest.fixture()
+def service_in_service_subtree(request):
+    princ = u'svc2/{0.host}@{0.realm}'.format(api.env)
+    rpcclient = api.Backend.rpcclient
+    was_connected = rpcclient.isconnected()
+
+    if not was_connected:
+        rpcclient.connect()
+
+    api.Command.service_add(princ)
+
+    def fin():
+        try:
+            api.Command.service_del(princ)
+        except Exception:
+            pass
+
+        try:
+            if not was_connected:
+                rpcclient.disconnect()
+        except Exception:
+            pass
+
+    request.addfinalizer(fin)
+    return princ
+
+
+@pytest.fixture(params=[service_in_kerberos_subtree,
+                        service_in_service_subtree])
+def service(request):
+    return request.param(request)
+
+
+@pytest.mark.skipif(
+    os.getuid() != 0, reason="kadmin.local is accesible only to root")
+class TestKadmin(object):
+    def assert_success(self, command, *args):
+        """
+        Since kadmin.local returns 0 also when internal errors occur, we have
+        to catch the command's stderr and check that it is empty
+        """
+        result = command(*args)
+        assert not result.error_output
```

Comment:
Warnings are also thrown in stderr (e.g. when no password policy was specified) but they are not tested in this case as they are part of setup. The problem here is, as Standa pointed out, that if the query fails, kardmin still returns 0, so testing for the presence of stderr is still the sanest way to assert command success. But I can re-implement the test to assert success by expecting certain messages in stdout if you wish.